flax/experimental/nnx/nnx/graph_utils.py

```diff
@@ -41,6 +44,45 @@
 
 NODE_TYPES: dict[type, 'NodeImpl[tp.Any, tp.Any, tp.Any]'] = {}
 
+class _HashById(tp.Hashable, tp.Generic[A]):
+  __slots__ = ('_value',)
+
+  def __init__(self, value: A):
+    self._value = value
+
+  @property
+  def value(self) -> A:
+    return self._value
+
+  def __hash__(self) -> int:
+    return id(self._value)
+
+  def __eq__(self, other: tp.Any) -> bool:
+    return isinstance(other, _HashById) and self._value is other._value
+
+
+class RefMap(tp.MutableMapping[A, B]):
```

Comment:
Consider adding a short docstring explaining the intent here, i.e. the keys are hashed based on object identity regardless of what `__hash__` for `A` actually does.